substrate/coins/pallet/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+  #[pallet::generate_deposit(fn deposit_event)]
+  pub enum Event<T: Config> {
+    Mint { address: SeraiAddress, balance: Balance },
+    Burn { address: SeraiAddress, balance: Balance, instruction: OutInstruction },
+    SriBurn { address: SeraiAddress, amount: Amount },
+    Transfer { from: SeraiAddress, to: SeraiAddress, balance: Balance },
+  }
+
+  #[pallet::pallet]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The amount of funds each account has.
+  #[pallet::storage]
+  #[pallet::getter(fn coins)]
+  pub type Coins<T: Config> = StorageDoubleMap<
+    _,
+    Blake2_128Concat,
+    T::AccountId,
+    Identity,
+    Coin,
+    SubstrateAmount,
+    OptionQuery,
+  >;
+
+  /// The total supply of each coin.
+  #[pallet::storage]
+  #[pallet::getter(fn supply)]
+  pub type Supply<T: Config> = StorageMap<_, Blake2_128Concat, Coin, SubstrateAmount, ValueQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+    fn build(&self) {
+      // initialize the supply of the coins
+      for c in COINS.iter() {
+        Supply::<T>::set(c, 0);
+      }
+
+      // initialize the genesis accounts
+      for (account, coin, amount) in self.accounts.iter() {
+        Coins::<T>::set(account, coin, Some(*amount));
+        let current_supply = Supply::<T>::get(coin);
+        Supply::<T>::set(coin, current_supply.checked_add(*amount).unwrap());
+      }
+    }
+  }
+
+  #[pallet::hooks]
+  impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+      // burn the fees collected previous block
+      let k1: T::AccountId = FEE_ACCOUNT.into();
+      let amount = Self::coins(k1, Coin::Serai).unwrap_or(0);
+      if amount != 0 {
+        // we can unwrap, we are not burning more then what we have.
+        Self::burn_internal(&FEE_ACCOUNT.into(), Coin::Serai, amount, None).unwrap();
+      }
+      Weight::zero() // TODO
+    }
+  }
+
+  impl<T: Config> Pallet<T> {
+    /// Returns the balance of a given account for `coin`.
+    pub fn balance(of: &T::AccountId, coin: Coin) -> SubstrateAmount {
+      Self::coins(of, coin).unwrap_or(0)
+    }
+
+    fn increase_supply(coin: Coin, by: SubstrateAmount) -> Result<(), Error<T>> {
+      // update the supply
+      let new_supply = Self::supply(coin).checked_add(by).ok_or(Error::<T>::AmountOverflowed)?;
+      Supply::<T>::set(coin, new_supply);
+      Ok(())
+    }
+
+    fn decrease_supply(coin: Coin, by: SubstrateAmount) -> Result<(), Error<T>> {
+      // update the supply
+      let new_supply = Self::supply(coin).checked_sub(by).ok_or(Error::<T>::AmountOverflowed)?;
+      Supply::<T>::set(coin, new_supply);
+      Ok(())
+    }
+
+    /// Mints amount at the given account, errors if amount overflows.
+    pub fn mint(at: &T::AccountId, coin: Coin, amount: SubstrateAmount) -> Result<(), Error<T>> {
+      // don't waste time if amount 0.
+      if amount == 0 {
+        return Ok(());
+      }
```

Comment:
Ironically enough, I think this check will waste more time than not having this check would. This'll be called by emissions, which should be non-zero, and InInstructions. The processor will drop anything that's dust (such as 0 transfers).

You're welcome to keep it as an idea of good practice though.